### ISSUE TITLE
IntoBytes for MaybeUninit<T>

### ISSUE THREAD

**Issue by kartva** [google/zerocopy#2172]
(not attaching the name of my project since it's still a prototype)

When building a zerocopy network protocol, I want to be able to have a function with the following signature:

```rs
pub unsafe fn send(&mut self, fill: impl FnOnce(&mut MaybeUninit<Payload>)) {
        let mem = &mut packet[packet_start_offset..];
        fill(MaybeUninit::<Payload>::mut_from_bytes(mem).unwrap());
        ...
}
```
so that users can just write to the uninit data and thus write to the underlying byte buffer.

Are there any reasons for why this isn't possible / desirable?

**Comment by kupiakos** [comment]
The reason this isn't directly possible is because one can pass this in for `fill`: `|x| *x = MaybeUninit::uninit()`, thus deinitializing previously initialized memory in `packet`.

You want something like ["out" references](https://docs.rs/uninit/0.6.2/uninit/out_ref/index.html):

```rust
use uninit::extension_traits::{AsOut, ManuallyDropMut};
use uninit::out_ref::Out;

pub unsafe fn send(&mut self, fill: impl FnOnce(Out<'_, Payload>)) {
        let payload = Payload::mut_from_bytes(&mut packet[packet_start_offset..]).unwrap();
        fill(MaybeUninit::<Payload>::mut_from_bytes(payload.manually_drop_mut().as_out()).unwrap());
        ...
}
```

**Comment by joshlf** [comment]
Yep, @kupiakos is right unfortunately. `T: IntoBytes` requires that all of `T`'s bytes are always initialized. `MaybeUninit` doesn't satisfy that requirement unless it is zero bytes long (see https://github.com/google/zerocopy/issues/123).
